compiler/generator_rust.c

```diff
@@ -0,0 +1,1347 @@
+}
+
+static void generate_sliceto(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->V[0] = p->name;
+    writef(g, "~M~V0 = env.slice_to(~V0);~N"
+              "~Mif ~V0 == ''{~N"
+              "~+~Mreturn false;~N~-~M}~N"
+            , p);
+}
+
+static void generate_address(struct generator * g, struct node * p) {
+
+    symbol * b = p->literalstring;
+    if (b != 0) {
+        write_literal_string(g, b);
+    } else {
+        write_varref(g, p->name);
+    }
+}
+
+static void generate_insert(struct generator * g, struct node * p, int style) {
+
+    int keep_c = style == c_attach;
+    write_comment(g, p);
+    if (p->mode == m_backward) keep_c = !keep_c;
+    if (keep_c) w(g, "~Mlet c = env.cursor;~N");
+    w(g, "~Mlet (bra, ket) = (env.cursor, env.cursor);~N");
+    writef(g, "~Menv.insert(bra, ket, ", p);
+    generate_address(g, p);
+    writef(g, ");~N", p);
+    if (keep_c) w(g, "~Menv.cursor = c;~N");
+}
+
+static void generate_assignfrom(struct generator * g, struct node * p) {
+
+    int keep_c = p->mode == m_forward; /* like 'attach' */
+
+    write_comment(g, p);
+    if (keep_c) writef(g, "~Mc = env.cursor;~N", p);
+    if (p->mode == m_forward) {
+        writef(g, "~Menv.insert(env.cursor, env.limit, ", p);
+    } else {
+        writef(g, "~Menv.insert(env.limit_backward, env.cursor, ", p);
+    }
+    generate_address(g, p);
+    writef(g, ")~N", p);
+    if (keep_c) w(g, "~Menv.cursor = c;~N");
+}
+
+
+static void generate_slicefrom(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    w(g, "~Mif !env.slice_from(");
+    generate_address(g, p);
+    writef(g, "){~N"
+              "~+~Mreturn false;~N~-~M}~N", p);
+}
+
+static void generate_setlimit(struct generator * g, struct node * p) {
+
+    struct str * savevar = vars_newname(g);
+    struct str * varname = vars_newname(g);
+    write_comment(g, p);
+    write_savecursor(g, p, savevar);
+    generate(g, p->left);
+
+    if (!g->unreachable) {
+        g->B[0] = str_data(varname);
+        if (p->mode == m_forward) {
+            w(g, "~Mlet ~B0 = env.limit - env.cursor;~N");
+            w(g, "~Menv.limit = env.cursor;~N");
+        } else {
+            w(g, "~Mlet ~B0 = env.limit_backward;~N");
+            w(g, "~Menv.limit_backward = env.cursor;~N");
+        }
+        write_restorecursor(g, p, savevar);
+
+        if (p->mode == m_forward) {
+            str_assign(g->failure_str, "env.limit += ");
+            str_append(g->failure_str, varname);
+            str_append_string(g->failure_str, ";");
+        } else {
+            str_assign(g->failure_str, "env.limit_backward = ");
+            str_append(g->failure_str, varname);
+            str_append_string(g->failure_str, ";");
+        }
+        
+        
+        generate(g, p->aux);
+
+        if (!g->unreachable) {
+            write_margin(g);
+            write_str(g, g->failure_str);
+            w(g, ";");
```

Comment:
This semicolon is redundant because both failure strings a few lines above already end with semicolons.